pykern/util.py

```diff
@@ -65,3 +65,40 @@ def _check_dev_files(root):
         # Don't run from an install directory
         r = pkio.py_path()
     return pkio.mkdir_parent(r.join(_DEFAULT_ROOT))
+
+
+def is_pure_text(bytes_data, chunk_size=None):
+    """Guesses if bytes_data is pure text by attempting to
+    utf-8 decode it. If decode fails, checks if chunk_size was
+    at the boundary of a valid truncated character
+
+    Args:
+        bytes_data (bytes): bytes data
+        chunk_size (int): size of bytes data
+
+    Returns:
+        bool: True if bytes_data is likely pure text, false if likely binary
+    """
+    if not chunk_size or len(bytes_data) < chunk_size:
+        return _decode_success(bytes_data, lambda: False)
+    b = bytes_data[:chunk_size]
+    return _decode_success(b, lambda: _char_was_truncated(b, bytes_data[chunk_size:]))
+
+
+def _char_was_truncated(bytes_data, extra_chars):
+    for i in range(len(extra_chars)):
+        try:
+            bytes_data += bytes([extra_chars[i]])
+            bytes_data.decode("utf-8", "strict")
+            return True
+        except UnicodeDecodeError:
+            continue
+    return False
+
+
+def _decode_success(data, decode_failure_callback):
+    try:
+        data.decode("utf-8", "strict")
+        return True
+    except UnicodeDecodeError:
+        return decode_failure_callback()
```

Comment:
I changed some names to be a bit more descriptive. "chunk" implies multiple chunks. "test_size" is suboptimal, but good enough. Possibly "probe_len".

Try to write less code, especially avoiding callbacks unless absolutely necessary.

Note the change in the test on test_size. If value is exactly test_size, just test it, you can't probe ahead anyway.

By going backwards, you avoid the read ahead problem. The test_size is a guess so we can go backwards, and it will be good enough.

```py
def is_pure_text(value, test_size=None):
    def _try(chunk):
        try:
            chunk.decode("utf-8", "strict")
            return True
        except UnicodeDecodeError:
            return False

    if not test_size or len(value) <= test_size:
        return _try(value)
    b = value[:chunk_size]
    # 4 is maximum length of a utf8 char so if a char
    # is truncated by test_size, we need to probe back
    # a bit to find the end of the char.
    for _ in range(4):
        if _try(b):
            return True
        if len(b) <= 1:
            return False
        b = b[:-1]
    return False
```